Replace `cloneDeep` for `.~/data/reducer.js` and apply new state mutation function to whole reducer

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #270

- Add `chain` and `set` functions to update state without mutating other references of state tree.
- Replace all `cloneDeep` and `getNextStateForShipping`.
- Replace the remaining reducer with the `chain` and `set` functions.
- Revert the temporary commit https://github.com/woocommerce/google-listings-and-ads/commit/ece1da7bb7fe7ae9a1e69db673874b0ad7a028c3 to check references of state tree for all test cases.

:warning: It depends on #1150.

### Detailed test instructions:

1. `npm run test:js:watch -- reducer`
2. Check if all test cases are passed.
3. Browse GLA pages to see if there are any unusual results or errors

### Changelog entry

> Fix - Replace `cloneDeep` within `.~/data/reducer.js` with functions that would not mutate other references of the state tree.

eason9487 committed:
Add a function to update state without mutating other references of state tree, and replace cloneDeep with it

eason9487 committed:
Replace getNextStateForShipping for reducer

eason9487 committed:
Replace the remaining reducer with the new setter function

eason9487 committed:
Revert "Comment out the reference checks of state tree that currently fails"

This reverts commit ece1da7bb7fe7ae9a1e69db673874b0ad7a028c3.

jorgemd24 commented on js/src/data/reducer.js:
```diff
@@ -37,29 +37,37 @@ const DEFAULT_STATE = {
-const getNextStateForShipping = ( state ) => {
+// Referenced and modified from https://github.com/lodash/lodash/issues/1696#issuecomment-328335502
+function chain( state, basePath = '' ) {
+	const nextState = clone( state );
```
Is it needed to use "clone" from lodash? We could use the spread operator `const nextState = {...state}` or Object.assign to create the nextState constant, allowing us to reduce the dependency of lodash library.

eason9487 replied:
Although the internal `customizer` still has some special handling for the data structure and usage of this reducer's state, the `chain` function was written based on generality, so I didn't assume that the `state` argument has to be this reducer's state, nor is it necessarily always the object type. For example.
- if there's a new reducer whose root node is an array.
- called with sub-nodes of the state tree. Such as the below could be replaced by `const newAdsCampaigns = set( state.ads_campaigns, idx, updatedCampaign );`
    https://github.com/woocommerce/google-listings-and-ads/blob/3c7ffb4ca267f8dfaa04ef56f65ce97b806cc50a/js/src/data/reducer.js#L202-L203

Or, we could
1. replace it with `const nextState = Array.isArray( state ) ? [ ...state ] : { ...state };`, but the test coverage report will report this line is uncovered. 
2. replace it with `const nextState = { ...state };` and add the array supporting once we needed.

What do you think?

jorgemd24 replied:
Thanks for the explanation! 😃

About the alternatives: 

1. Not ideal if this line is uncovered... and I think the code looks cleaner as it is now (with the clone function). 
2. It is the easiest one but that means that we will need to go back at some point to tweak it... so it is not ideal.

What about this: 💭

`const nextState = Object.assign( state.constructor(), state );`

Depending on the instance that we receive we create a new instance. That means that we can work with arrays and objects seamless.

eason9487 replied:
Thanks for the suggestion! Interesting idea! 😆  I replaced that `clone` with the proposed one by https://github.com/woocommerce/google-listings-and-ads/pull/1151/commits/51a50200ed312317fcf09a9a781f5b22b4b3dfd4.

jorgemd24 replied:
Thanks! 😃

jorgemd24 commented on js/src/data/reducer.js:
```diff
@@ -266,103 +202,67 @@ const reducer = ( state = DEFAULT_STATE, action ) => {
-			};
-
-			newState.mc_issues.issues.splice(
+			const issues = state.mc_issues?.issues?.slice() || [];
```
I am assuming that `state.mc_issues.issues` can only be an array, false or undefined and therefore it is unnecessary to check if the issues property is an array before we call the method slice()

eason9487 replied:
Nice catch! Thanks! Removed by https://github.com/woocommerce/google-listings-and-ads/pull/1151/commits/7c511e78dcf800460b046362a381cf59e540631d

jorgemd24 commented on js/src/data/reducer.js:
```diff
@@ -116,141 +117,76 @@ const reducer = ( state = DEFAULT_STATE, action ) => {
-			const newState = cloneDeep( state );
-			newState.mc.accounts.ads_billing_status = billingStatus;
-			return newState;
+			const path = 'mc.accounts.ads_billing_status';
+			return set( state, path, action.billingStatus );
```
💅   we could set the path straight into the set function arguments, so it follows the same structure as the others. 

`return set( state,'mc.accounts.ads_billing_status', action.billingStatus );`

eason9487 replied:
Since `prettier` will ask for the below format, I tried to maintain visual consistency from the other perspective, so I wrote it in two lines.

```js
return set(
	state,
	'mc.accounts.ads_billing_status',
	action.billingStatus
);
```

Changed by https://github.com/woocommerce/google-listings-and-ads/pull/1151/commits/2a8f6af673e3b5f45f248d221c65d8032fb2f2f6

jorgemd24 reviewed (commented):
I see that this was a hard job and it wasn’t an easy task to update the reducer. 

The functions set & chain allowed to remove a lot of duplicated code and made the code easier to read it.

Regarding the code, I left some comments about the clone function and other minor things on the code. 

What I tested: 

✅  I run the reducers test and all are 🟢
✅  All tests are 🟢
✅  I browsed using my local environment (PHP 7.4, WordPress 5.8.2 & WC 5.9) and everything works fine. 

Thanks, it was an amazing job! 💪

eason9487 committed:
Call `set` directly without declaring other variables when simply updating state in the reducer.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1151#discussion_r767301208

eason9487 committed:
Remove an unnecessary optional chaining operator.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1151#discussion_r767300146

eason9487 committed:
Remove a TODO comment after confirming that it works as usual.

Ref https://github.com/woocommerce/google-listings-and-ads/pull/277#discussion_r586112399

eason9487 commented:
💡 
I added an [additional commit](https://github.com/woocommerce/google-listings-and-ads/pull/1151/commits/6b85d522b59d23f6b3b2038be8f5e83d0c8ef414) to remove a relevant TODO comment since I checked it works well as usual. More details can be found in [this comment](https://github.com/woocommerce/google-listings-and-ads/pull/277#discussion_r586112399).

eason9487 committed:
Use another way to shallow copy an object or array of passed-in `state` for the `chain` function in reducer.

Co-authored-by: Jorge Mahecha <jorge.mahecha@a8c.com>

jorgemd24 commented on js/src/data/reducer.js:
```diff
@@ -37,29 +37,37 @@ const DEFAULT_STATE = {
+		if ( value === null || value === undefined ) {
+			return {};
+		}
+		return clone( value );
```
What do you think if we replace this `clone` with `return Object.assign( value.constructor(), value )` so the only thing depending on lodash (in this file) is the function setWith?

eason9487 replied:
Although we will almost not have the following usage to be committed to codebase, it is inevitable that the wrong `path` might be typed when developing, and the returned results would be more confusing than the current implementation to figure out how they are generated.

```js
function chain( state, basePath = '' ) {
	const nextState = Object.assign( state.constructor(), state );
	const customizer = ( value ) => {
		if ( value === null || value === undefined ) {
			return {};
		}
		return Object.assign( value.constructor(), value );
	};

	return {
		set( path, value ) {
			const fullPath = basePath ? `${ basePath }.${ path }` : path;
			setWith( nextState, fullPath, value, customizer );
			return this;
		},
		end: () => nextState,
	};
}

function set( state, path, value ) {
	return chain( state ).set( path, value ).end();
}

set( { foo: 123 }, 'foo.bar', 456 );
// Got
// {
//   foo: [Number: 0] {
//     bar: 456
//   }
// }
// => `foo` property is a number object with value 0,
//    and that number object has a `bar` property with 456.
//    And it even can access the 456 value via newState.foo.bar path.

set( { foo: 'Hello' }, 'foo.bar', 'WordPress' );
// Got
// {
//   foo: [String: ''] {
//     '0': 'H',
//     '1': 'e',
//     '2': 'l',
//     '3': 'l',
//     '4': 'o',
//     bar: 'WordPress'
//   }
// }
// => `foo` property is a string object with enumerated keys/values of 'Hello',
//    and that string object has a `bar` property with 'WordPress'
```

The current implementation is easier to find that the wrong path was typed. Surely this is a relative edge case, so I think it won't be too many problems even if we use the proposed method instead. 

However, Sorry I didn't prepare a summary of the historical source of the solution. 🙇‍♂️  There is a lot of context to this PR.

Reducing dependency on lodash or gradually phasing out lodash is indeed a long-term goal for some teams. While we haven't explicitly decided whether to follow those principles, in the short-term goals discussed in #270, we were tending to resolve it with the effective solution found from lodash GitHub and less of our custom implementations. Some relevant comments:

- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-863080575
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-863169462
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-863412160
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-864708881
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-864799355
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-883266968

The goal of this PR is to solve the #270 problem based on the previous conclusion. If the concern is to reduce the reliance on lodash, I'd be happy to open another issue or p2 to discuss it because currently it's a bit out of the scope of this PR's purpose. 🙏  So, if there're no concerns other than those mentioned above, I would prefer to keep the lodash solution as it is and only adjusts the parts we can't avoid.

jorgemd24 replied:
I see the side effects that would cause if the path is typed wrongly when developing and I agree with you that for now, it is better to keep the current implementation until a decision is taken about lodash dependency.

jorgemd24 reviewed (approved):
As we mentioned before, all the minor issues have been resolved or discussed. 👍

This PR LGTM 👍
